### PR TITLE
tests: Check that golden tokens add up to the input

### LINKED ISSUE #2578: Check that lexers produce the same output as the input
On https://github.com/pygments/pygments/pull/2531, I just noticed missing groups in rules using `bygroups()`, making the lexer lose part of the input. (I've seen it on other PRs too.)

We do have this test in `test_basic_api.py`:

```python
    assert txt == test_content, "%s lexer roundtrip failed: %r != %r" % \
        (cls.name, test_content, txt)
```

but it only runs on random input.

Let's at least do that when highlighting the tests in `tests/snippets/` and `tests/examplefiles/`.

### PR DESCRIPTION
**Pull request by jeanas**
Fixes #2578

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -59,12 +66,18 @@ def _prunetraceback(self, excinfo):
...
-            rel_path = self._test_file_rel_path()
-            message = (
-                'The tokens produced by the "{}" lexer differ from the '
-                'expected ones in the file "{}".\n'
-                'Run `pytest {} --update-goldens` to update it.'
-            ).format(self.lexer, rel_path, Path(*rel_path.parts[:2]))
+            if self.config.getoption('--update-goldens'):
+                message = (
+                    f'The tokens produced by the "{self.lexer}" lexer '
+                    'does not add up to the input.'
```
Comment (birkenfeld) [inline tests/conftest.py:72]:
```suggestion
                    'do not add up to the input.'
```

File: tests/conftest.py [hunk 2]
```diff
@@ -59,12 +66,18 @@ def _prunetraceback(self, excinfo):
...
+                )
+            else:
+                rel_path = self._test_file_rel_path()
+                message = (
+                    'The tokens produced by the "{}" lexer differ from the '
+                    'expected ones in the file "{}".\n'
+                    'Run `pytest {} --update-goldens` to update it.'
```
Comment (birkenfeld) [inline tests/conftest.py:79]: Should this refer to tox?

File: pygments/lexer.py [hunk 3]
```diff
@@ -255,6 +255,8 @@ def get_tokens(self, text, unfiltered=False):
         if self.ensurenl and not text.endswith('\n'):
             text += '\n'
 
+        self._input_for_tests = text
```
Comment (birkenfeld) [inline pygments/lexer.py:258]: This looks kludgy, is there a way to avoid storing the input indefinitely in production?

**Review by birkenfeld — APPROVED** [review]

**Comment by Anteru** [comment]
Nice!

**Pull request merged**